src/safemode_ui.cpp

```diff
@@ -0,0 +1,740 @@
+
+        // Clear the lines
+        for( int i = 0; i < content_height; i++ ) {
+            for( int j = 0; j < 79; j++ ) {
+                mvwputch( w, i, j, c_black, ' ' );
+            }
+
+            for( auto &pos : column_pos ) {
+                mvwputch( w, i, pos.second, c_ltgray, LINE_XOXO );
+            }
+        }
+
+        auto &current_tab = ( tab == GLOBAL_TAB ) ? global_rules : character_rules;
+        const bool current_tab_non_empty = !current_tab.empty();
+
+        if( tab == CHARACTER_TAB && g->u.name == "" ) {
+            character_rules.clear();
+            mvwprintz( w, 8, 15, c_white, _( "Please load a character first to use this page!" ) );
+        }
+
+        draw_scrollbar( w_border, line, content_height, current_tab.size(), 5 );
+        wrefresh( w_border );
+
+        calcStartPos( start_pos, line, content_height, current_tab.size() );
+
+        // display safemode
+        for( int i = start_pos; i < ( int )current_tab.size(); i++ ) {
+            if( i >= start_pos &&
+                i < start_pos + ( ( content_height > ( int )current_tab.size() ) ?
+                                  ( int )current_tab.size() : content_height ) ) {
+
+                auto rule = current_tab[i];
+
+                nc_color line_color = ( rule.active ) ? c_white : c_ltgray;
+
+                temp_stream.str( "" );
+                temp_stream << i + 1;
+                mvwprintz( w, i - start_pos, 1, line_color, "%s", temp_stream.str().c_str() );
+                mvwprintz( w, i - start_pos, 5, c_yellow, ( line == i ) ? ">> " : "   " );
+
+                auto draw_column = [&]( Columns column_in, std::string text_in ) {
+                    mvwprintz( w, i - start_pos, column_pos[column_in] + 2,
+                               ( line == i && column == column_in ) ? hilite( line_color ) : line_color,
+                               "%s", text_in.c_str()
+                             );
+                };
+
+                draw_column( COLUMN_RULE, ( rule.rule == "" ) ? _( "<empty rule>" ) : rule.rule );
+                draw_column( COLUMN_ATTITUDE, Creature::get_attitude_ui_data( rule.attitude ).first );
+                draw_column( COLUMN_PROXIMITY, ( !rule.exclude ) ? to_string( rule.proximity ).c_str() : "---" );
+                draw_column( COLUMN_INCLUDE_EXCLUDE, ( rule.exclude ) ?  _( "Exclude" ) : _( "Include" ) );
+            }
+        }
+
+        wrefresh( w );
+
+        const std::string action = ctxt.handle_input();
+
+        if( action == "NEXT_TAB" ) {
+            tab++;
+            if( tab >= MAX_TAB ) {
+                tab = 0;
+                line = 0;
+            }
+        } else if( action == "PREV_TAB" ) {
+            tab--;
+            if( tab < 0 ) {
+                tab = MAX_TAB - 1;
+                line = 0;
+            }
+        } else if( action == "QUIT" ) {
+            break;
+        } else if( tab == CHARACTER_TAB && g->u.name.empty() ) {
```

Comment:
Now I'm wondering why this tests for `empty()`, while other code in this PR uses `== ""`. Have the same behavior, but using `empty()` is generic and states the intend clearer.

Anyway, this is inconsistent. (Through the problem is not just here.)